cargo-miri/bin.rs

```diff
@@ -652,84 +698,93 @@ fn phase_cargo_runner(binary: &Path, binary_args: env::Args) {
+            let json_flag = "--json";
+            while let Some(arg) = args.next() {
+                if arg == extern_flag {
+                    cmd.arg(extern_flag); // always forward flag, but adjust filename
+                    // `--extern` is always passed as a separate argument by cargo.
+                    let next_arg = args.next().expect("`--extern` should be followed by a filename");
+                    if let Some(next_lib) = next_arg.strip_suffix(".rlib") {
+                        // If this is an rlib, make it an rmeta.
+                        cmd.arg(format!("{}.rmeta", next_lib));
+                    } else {
+                        // Some other extern file (e.g., a `.so`). Forward unchanged.
+                        cmd.arg(next_arg);
+                    }
+                } else if arg.starts_with(error_format_flag) {
+                    let suffix = &arg[error_format_flag.len()..];
+                    assert!(suffix.starts_with('='));
+                    // Drop this argument.
+                } else if arg.starts_with(json_flag) {
+                    let suffix = &arg[json_flag.len()..];
+                    assert!(suffix.starts_with('='));
+                    // Drop this argument.
+                } else {
+                    cmd.arg(arg);
+                }
+            }
+            // Set sysroot.
+            let sysroot =
+                env::var_os("MIRI_SYSROOT").expect("the wrapper should have set MIRI_SYSROOT");
+            cmd.arg("--sysroot");
+            cmd.arg(sysroot);
+            // Respect `MIRIFLAGS`.
+            if let Ok(a) = env::var("MIRIFLAGS") {
+                // This code is taken from `RUSTFLAGS` handling in cargo.
+                let args = a
+                    .split(' ')
+                    .map(str::trim)
+                    .filter(|s| !s.is_empty())
+                    .map(str::to_string);
+                cmd.args(args);
+            }
 
-    // Make sure we use the build-time working directory for interpreting Miri/rustc arguments.
-    // But then we need to switch to the run-time one, which we instruct Miri do do by setting `MIRI_CWD`.
-    cmd.current_dir(info.current_dir);
-    cmd.env("MIRI_CWD", env::current_dir().unwrap());
+            // Then pass binary arguments.
+            cmd.arg("--");
+            cmd.args(binary_args);
 
-    // Run it.
-    if verbose {
-        eprintln!("[cargo-miri runner] {:?}", cmd);
+            // Make sure we use the build-time working directory for interpreting Miri/rustc arguments.
+            // But then we need to switch to the run-time one, which we instruct Miri do do by setting `MIRI_CWD`.
+            cmd.current_dir(current_dir);
+            cmd.env("MIRI_CWD", env::current_dir().unwrap());
+
+            // Run it.
+            if verbose {
+                eprintln!("[cargo-miri runner] {:?}", cmd);
+            }
+            exec(cmd)
+        }
+        CrateRunInfo::Skip => {
```

Comment:
To avoid rightward drift, it'd be good to make this code more like
```
let info_inner = match info {
  CrateRunInfo::RunWith(inner) => inner,
  CrateRunInfo::Skip ==> {
    eprintln!(...);
    return;
  }
};
```
Then the control flow here doesn't test quite as deeply.
```

Reply:
`CrateRunInfo::RunWith` is a struct variant, so changing to something like that requires either creating a tuple with tree elements or packing fields of that variant in another struct. I'm not sure which is more desirable.

Reply:
I think I slightly prefer having a new type for this.